### PR TITLE
bpo-19610: Warn if distutils is provided something other than a list to some fields

### PR DESCRIPTION
**Pull request by nascheme**
The commit of dcaed6b2d95 causes some 3rd party packages to fail to install, due to them passing a tuple for meta data fields.  Rather than raise TypeError, generate a RuntimeWarning and convert the value using list().

https://bugs.python.org/issue19610

### CODE REVIEW

File: Lib/distutils/tests/test_dist.py [hunk 1]
```diff
@@ -354,8 +354,11 @@ def test_classifier_invalid_type(self):
         attrs = {'name': 'Boa', 'version': '3.0',
                  'classifiers': ('Programming Language :: Python :: 3',)}
         msg = "'classifiers' should be a 'list', not 'tuple'"
-        with self.assertRaises(TypeError, msg=msg):
-            Distribution(attrs)
+        with self.assertWarns(RuntimeWarning) as warns:
```
Comment (serhiy-storchaka) [inline Lib/distutils/tests/test_dist.py]: Would is better to use `assertWarnsRegex()` here and in other tests. Otherwise `msg` isn't used.

File: Lib/distutils/tests/test_dist.py [hunk 2]
```diff
@@ -354,8 +354,11 @@ def test_classifier_invalid_type(self):
...
+            d = Distribution(attrs)
+            self.assertIsInstance(d.metadata.classifiers, list)
```
Comment (serhiy-storchaka) [inline Lib/distutils/tests/test_dist.py]: Would be better to move these checks outside of the `with` block. Otherwise it isn't clear what operation raises a RuntimeWarning.

**Review by serhiy-storchaka — COMMENTED** [review]
Needed to update `Doc/distutils/apiref.rst` too.

**Review by serhiy-storchaka — APPROVED** [review]

File: Lib/distutils/dist.py [hunk 3]
```diff
@@ -9,10 +9,7 @@
 import re
 from email import message_from_file
 
-try:
-    import warnings
-except ImportError:
-    warnings = None
+import warnings
```
Comment (merwok) [inline Lib/distutils/dist.py]: It’s possible that the import was how it was because distutils is used to build CPython’s own extension modules.  In doubt, let’s not change this.

**Review by merwok — CHANGES_REQUESTED** [review]
Thanks for the patch.  Tests look good; left comment about conditional use of warnings.

File: Lib/distutils/dist.py (hunk 3 above)
Comment (merwok) [reply Lib/distutils/dist.py]: Note that distutils also has its own PEP-282-style logging system, which may or may not easier to use here.  The output is controlled by verbose/quiet options passed to setup.py rather than the interpreter warning configuration system, so testing with pip would be needed to make sure that package authors would see the messages by default.

Comment (nascheme) [reply Lib/distutils/dist.py]: The try/except dates back to Python 1.5 + 2.0 days.  It was added because those versions didn't have 'warnings' and distutils was targeting those versions as well.  I think removing it is safe but I will revert as it is a separate change from this issue.

Using the logging system sounds like a good idea.  I'll look into that.

**Comment by serhiy-storchaka** [comment]
While we a here could you please add `value = list(value)` in `set_requires()` and `set_obsoletes()`. They already work with arbitrary iterables, but this can have unexpected effect in `upload.py` and `register.py`.

File: Lib/distutils/dist.py (hunk 3 above)
Comment (nascheme) [reply Lib/distutils/dist.py]: I have made the requested changes; please review again.

I tweaked the wording of the warning message, not sure it is the best.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by nascheme** [comment]
I have made the requested changes; please review again.

File: Lib/distutils/dist.py [hunk 4]
```diff
@@ -257,10 +269,7 @@ def __init__(self, attrs=None):
                     setattr(self, key, val)
                 else:
                     msg = "Unknown distribution option: %s" % repr(key)
-                    if warnings is not None:
-                        warnings.warn(msg)
-                    else:
-                        sys.stderr.write(msg + "\n")
+                    warnings.warn(msg)
```
Comment (merwok) [inline Lib/distutils/dist.py:274]: I thought you wanted to move this in a separate PR?

File: Lib/distutils/dist.py [hunk 5]
```diff
@@ -26,6 +26,18 @@
...
 
+def _ensure_list(value, fieldname):
+    if isinstance(value, str):
+        # a string containing comma separated values is okay.  It will
+        # be converted to a list by Distribution.finalize_options().
+        pass
+    elif not isinstance(value, list):
+        # passing a tuple or an iterator perhaps, warn and convert
+        typename = type(value).__name__
+        msg = f"Warning: '{fieldname}' should be a list, got type '{typename}'"
```
Comment (merwok) [inline Lib/distutils/dist.py:38]: Is the prefix `Warning` redundant?  Otherwise message seems good.

File: Lib/distutils/dist.py [hunk 6]
```diff
@@ -26,6 +26,18 @@
 # to look for a Python module named after the command.
 command_re = re.compile(r'^[a-zA-Z]([a-zA-Z0-9_]*)$')
 
+def _ensure_list(value, fieldname):
```
Comment (merwok) [inline Lib/distutils/dist.py]: Minor: would you mind keeping two blank lines before and after this function?

File: Lib/distutils/tests/test_dist.py [hunk 7]
```diff
@@ -11,7 +11,8 @@
 from distutils.dist import Distribution, fix_help_options, DistributionMetadata
 from distutils.cmd import Command
 
-from test.support import TESTFN, captured_stdout, run_unittest
+from test.support import TESTFN, captured_stdout, captured_stderr, \
```
Comment (berkerpeksag) [inline Lib/distutils/tests/test_dist.py]: Style nit: Replacing `\` with

```py
from test.support import (
    TESTFN, captured_stdout, captured_stderr, run_unittest,
)
```

may be better.

File: Lib/distutils/dist.py [hunk 8]
```diff
@@ -1231,7 +1223,7 @@ def set_requires(self, value):
         import distutils.versionpredicate
         for v in value:
             distutils.versionpredicate.VersionPredicate(v)
-        self.requires = value
+        self.requires = list(value)
```
Comment (berkerpeksag) [inline Lib/distutils/dist.py:1228]: Should we add tests for the requires and obsolotes fields similar to what you already added for the classifiers field?:

```py
attrs = {...,
         'classifiers': ('Programming Language :: Python :: 3',)}
self.assertIsInstance(d.metadata.classifiers, list)
self.assertEqual(d.metadata.classifiers,
                 list(attrs['classifiers']))
```

If I don't miss something, the line `list(value)` is not covered by the current tests:

https://github.com/python/cpython/blob/bce095d9b0595a528a95bae3f764ee195a0c0d63/Lib/distutils/tests/test_dist.py#L301-L315
https://github.com/python/cpython/blob/bce095d9b0595a528a95bae3f764ee195a0c0d63/Lib/distutils/tests/test_dist.py#L323-L337

File: Lib/distutils/dist.py (hunk 5 above)
Comment (berkerpeksag) [reply Lib/distutils/dist.py:38]: Another style nit: Can't we use `{fieldname!r}` instead of `'{fieldname}'`?

File: Lib/distutils/tests/test_dist.py [hunk 9]
```diff
@@ -381,9 +389,12 @@ def test_platforms(self):
     def test_platforms_invalid_types(self):
         attrs = {'name': 'Monty', 'version': '1.0',
                  'platforms': ('GNU/Linux', 'Some Evil Platform')}
-        msg = "'platforms' should be a 'list', not 'tuple'"
-        with self.assertRaises(TypeError, msg=msg):
-            Distribution(attrs)
+        msg = "should be a list"
```
Comment (berkerpeksag) [inline Lib/distutils/tests/test_dist.py]: Style nit: I think we can now inline this in line 395:

```py
self.assertIn(msg, error.getvalue())
```

**Review by berkerpeksag — COMMENTED** [review]
This looks pretty good to me, thank you! I just left some comments on code style.
